### PR TITLE
[libssh] Add support for Android

### PR DESCRIPTION
**Pull request by ahmedyarub**
**Describe the pull request**
Allows using VCPKG toolchain + Android toolchain. Otherwise, we would not be able to use Android toolchain together with VCPKG building libssh. After this, SDK building should work on VCPKG.
This change can be easily extended to enable building with VCPKG + iOS/Wasm or any other cross-platform build.

*Check all that applies:*
- [x] Did a review by yourself.
- [ ] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [ ] Checked if this PR is a breaking (APIs have been changed) change.
- [x] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [x] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [ ] Linux
- [ ] Windows
- [x] Android
- [ ] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

****

### CODE REVIEW

File: port_versions/baseline.json [hunk 1]
```diff
@@ -3338,7 +3338,7 @@
     },
     "libssh": {
       "baseline": "0.9.5",
-      "port-version": 0
+      "port-version": 2
```
Comment (NancyLi1013) [inline port_versions/baseline.json:3341]:
```suggestion
      "port-version": 1
```

File: port_versions/l-/libssh.json [hunk 2]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "3377cffea59634bbc1bdef43c2ad618e7165d88a",
+      "version-string": "0.9.5",
+      "port-version": 2
```
Comment (NancyLi1013) [inline port_versions/l-/libssh.json:6]:
```suggestion
      "port-version": 1
```

File: ports/libssh/portfile.cmake [hunk 3]
```diff
@@ -19,10 +19,20 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
...
+			"-DWITH_SERVER=FALSE"
+			"-DWITH_PCAP=FALSE"
+			)
+endif ()
+
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
     OPTIONS
+		${EXTRA_ARGS}
```
Comment (NancyLi1013) [inline ports/libssh/portfile.cmake:35]:
```suggestion
    OPTIONS
        ${EXTRA_ARGS}
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by ahmedyarub** [comment]
Requested changes made

File: ports/libssh/CONTROL [hunk 4]
```diff
@@ -1,5 +1,6 @@
 Source: libssh
 Version: 0.9.5
+Port-Version: 2
```
Comment (NancyLi1013) [inline ports/libssh/CONTROL:3]:
```suggestion
Port-Version: 1
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Comment by ahmedyarub** [comment]
@NancyLi1013 still waiting for my response?

**Comment by NancyLi1013** [comment]
x86-windows still failed. I will check this later.

**Comment by NancyLi1013** [comment]
For the files in _port_versions_ folder, it would be better to update via `vcpkg x-add-version libssh` instead of updating them manually.

**Comment by ahmedyarub** [comment]
> vcpkg x-add-version libssh

What about now?

**Comment by NancyLi1013** [comment]
Have you tested this triplet on your machine? @ahmedyarub

**Comment by ahmedyarub** [comment]
Yes, I did. I can even post a video if you want.
    On Monday, January 18, 2021, 10:44:48 PM GMT-3, NancyLi1013 <notifications@github.com> wrote:  
 
 


Have you tested this triplet on your machine? @ahmedyarub

**Comment by NancyLi1013** [comment]
Thanks for your feedback @ahmedyarub. It's unnecessary to provide the video. 

LGTM now, thanks for your PR.

File: ports/libssh/portfile.cmake [hunk 5]
```diff
@@ -19,10 +19,20 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
     zlib    WITH_ZLIB
 )
 
+if (VCPKG_TARGET_IS_ANDROID)
+	set(EXTRA_ARGS "-DTARGET_ARCH=ANDROID"
+			"-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=${VCPKG_CHAINLOAD_TOOLCHAIN_FILE}"
```
Comment (dan-shaw) [inline ports/libssh/portfile.cmake:24]: Can we remove this argument? This should be added when invoking CMake instead.

Comment (ahmedyarub) [reply ports/libssh/portfile.cmake:24]: but vcpkg_configure_cmake() does invoke CMake. Is there any specific place to put this argument that I'm not aware of?

Comment (ras0219-msft) [reply ports/libssh/portfile.cmake:24]: `vcpkg_configure_cmake()` already adds this option automatically here:

https://github.com/microsoft/vcpkg/blob/b41a46f3d758032944f5dbe2ed58d0d1e5e7dc30/scripts/cmake/vcpkg_configure_cmake.cmake#L231-L232

File: ports/libssh/portfile.cmake [hunk 6]
```diff
@@ -19,10 +19,20 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
     zlib    WITH_ZLIB
 )
 
+if (VCPKG_TARGET_IS_ANDROID)
+	set(EXTRA_ARGS "-DTARGET_ARCH=ANDROID"
```
Comment (ras0219-msft) [inline ports/libssh/portfile.cmake:23]: What consumes this `TARGET_ARCH` setting? If this isn't specifically a setting recognized by libssh, it almost certainly shouldn't be in this file.

File: ports/libssh/portfile.cmake [hunk 7]
```diff
@@ -19,10 +19,20 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
...
+			"-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=${VCPKG_CHAINLOAD_TOOLCHAIN_FILE}"
+			"-DWITH_MBEDTLS=TRUE"
```
Comment (ras0219-msft) [inline ports/libssh/portfile.cmake:25]: This is already controlled by the feature setting above on line 18; if the user wants mbedtls they should require `libssh[mbedtls]`.

If mbedtls is mandatory for android, you can add
```
Build-Depends: libssh[core,mbedtls] (android)
```
to the top paragraph.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

**Comment by ras0219-msft** [comment]
Thanks for all the android PRs!

**Comment by ahmedalnuaimi** [comment]
> Thanks for all the android PRs!

You are welcome but please hold on merging some major factoring is coming.

File: ports/libssh/portfile.cmake (hunk 6 above)
Comment (ahmedyarub) [reply ports/libssh/portfile.cmake:23]: Looks like I got things mixed up
https://stackoverflow.com/a/41395408/7669414
Removing...

Comment (ahmedyarub) [reply ports/libssh/portfile.cmake]: Fixed

Comment (ahmedalnuaimi) [reply ports/libssh/portfile.cmake]: Sorry @ras0219 , this parameter is available here https://github.com/aws/aws-sdk-cpp/blob/master/CMakeLists.txt#L165 which is related to aws-sdk-cpp and not libssh. Removed both from libssh and mbedtls and kept in aws-sdk-cpp.

File: ports/libssh/portfile.cmake (hunk 5 above)
Comment (ahmedyarub) [reply ports/libssh/portfile.cmake]: Fixed and tested according to https://vcpkg.readthedocs.io/en/latest/users/android/
Thank you very much!

**Comment by ahmedyarub** [comment]
Fixed according to https://vcpkg.readthedocs.io/en/latest/users/android/

File: ports/libssh/CONTROL [hunk 8]
```diff
@@ -10,7 +12,7 @@ Build-Depends: libssh[mbedtls]
 Description: Default crypto backend
 
 Feature: mbedtls
-Build-Depends: mbedtls[pthreads]
+Build-Depends: mbedtls[pthreads] (!android)
```
Comment (ras0219) [inline ports/libssh/CONTROL:15]:
```suggestion
Build-Depends: mbedtls[pthreads] (!android), mbedtls[core]
```
The qualifier `(!android)` applies to the entire dependency, not just the `[pthreads]` feature

File: ports/libssh/portfile.cmake [hunk 9]
```diff
@@ -19,10 +19,18 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
     zlib    WITH_ZLIB
 )
 
+if (VCPKG_TARGET_IS_ANDROID)
+	set(EXTRA_ARGS "-DWITH_MBEDTLS=TRUE"
+			"-DWITH_SERVER=FALSE"
+			"-DWITH_PCAP=FALSE"
+			)
```
Comment (ras0219) [inline ports/libssh/portfile.cmake:26]:
```suggestion
	set(EXTRA_ARGS "-DWITH_SERVER=FALSE"
			"-DWITH_PCAP=FALSE"
			)
```

`vcpkg_check_features()` will already add `-DWITH_MBEDTLS` based on the feature setting.

**Review by ras0219 — CHANGES_REQUESTED** [review]

File: ports/libssh/portfile.cmake (hunk 9 above)
Comment (ahmedyarub) [reply ports/libssh/portfile.cmake]: Fixed

**Comment by ahmedyarub** [comment]
I can't understand why these tests are failing. They are not related at all to my changes.

**Comment by NancyLi1013** [comment]
> I can't understand why these tests are failing. They are not related at all to my changes.

The failures are caused by our CI test environment. Now All checks have passed after rerun failed job.

**Comment by ahmedyarub** [comment]
> 
> 
> > I can't understand why these tests are failing. They are not related at all to my changes.
> 
> The failures are caused by our CI test environment. Now All checks have passed after rerun failed job.

So can it be merged?

**Pull request merged**